core/arch/aarchxx/mangle.c

```diff
@@ -2922,7 +2954,573 @@ mangle_icache_op(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+            /* We have no CBNZ so we need to preserve the flags. */
+            dr_save_arith_flags_to_reg(dcontext, ilist, instr, scratch);
+            PRE(ilist, instr,
+                instr_create_save_to_tls(dcontext, scratch, TLS_LDSTEX_FLAGS_SLOT));
+        }
+#endif
+        if (is_pair) {
+            scratch2 = pick_scratch_reg(dcontext, instr, swap_reg, scratch, DR_REG_NULL,
+                                        false, &slot2, &should_restore2);
+            scratch3 = pick_scratch_reg(dcontext, instr, swap_reg, scratch, scratch2,
+                                        false, &slot3, &should_restore3);
+            insert_save_to_tls_if_necessary(dcontext, ilist, instr, scratch2, slot2);
+            insert_save_to_tls_if_necessary(dcontext, ilist, instr, scratch3, slot3);
+        }
+        /* Compare address, arranging op_res to show failure on mismatch.
+         * XXX i#400: It is possible that the store could fault and the app could
+         * examine ops_res in the handler: i.e., it's not truly dead.  We do not
+         * account for that here.
+         */
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch, TLS_LDSTEX_ADDR_SLOT));
+        insert_compare_and_jump_not_equal(dcontext, ilist, instr, op_res,
+                                          opnd_create_reg(reg_base),
+                                          opnd_create_reg(scratch), no_match);
+
+        /* Compare size, arranging op_res to show failure on mismatch. */
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch, TLS_LDSTEX_SIZE_SLOT));
+        insert_compare_and_jump_not_equal(
+            dcontext, ilist, instr, op_res, opnd_create_reg(scratch),
+            OPND_CREATE_INT(reg_get_size(reg_st_val)), no_match);
+    } else {
+        /* We aren't saving the flags so we can only handle Thumb mode with CBNZ. */
+        IF_ARM(ASSERT(is_cbnz_available(dcontext, reg_res)));
+    }
+
+    /* Perform the compare-and-swap. */
+    if (!ldex_in_same_block) {
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch, TLS_LDSTEX_VALUE_SLOT));
+        reg_orig_ld_val = scratch;
+        if (is_pair) {
+            PRE(ilist, instr,
+                instr_create_restore_from_tls(dcontext, scratch2,
+                                              TLS_LDSTEX_VALUE2_SLOT));
+            reg_orig_ld_val2 = scratch2;
+        }
+    }
+    reg_id_t reg_new_ld_val;
+    reg_id_t reg_new_ld_val2 = is_pair ? scratch3 : DR_REG_NULL;
+    PRE(ilist, instr,
+        create_ldax_from_stex(dcontext, instr, &reg_new_ld_val, reg_new_ld_val2));
+    reg_new_ld_val = reg_to_pointer_sized(reg_new_ld_val);
+    insert_compare_and_jump_not_equal(dcontext, ilist, instr, op_res,
+                                      opnd_create_reg(reg_new_ld_val),
+                                      opnd_create_reg(reg_orig_ld_val), no_match);
+    if (is_pair) {
```

Comment:
Should we assert that `load_is_pair == store_is_pair`? This should really be true mostly, but maybe we can add a debug check.

Reply:
The app is free to not have both be a pair, with a failing monitor in that case.  The size check will fail if the app has one a pair and the other not.

Reply:
> The size check will fail if the app has one a pair and the other not.

I agree that DR will cause an infinite loop.


> The app is free to not have both be a pair, with a failing monitor in that case.

Do you mean a sequence like the following will be an infinite loop without DR too?

```
label:
ldxp x0, x1, [x2]
...
stx w3, x0, [x2]
cbnz w3, label
```

Reply:
Yes, that is an infinite loop natively (with the fix s/stx/stxr/).

I tested it to make sure DR did the right thing: and it doesn't!  There are bugs in its size checks which make it think the sizes match.  Grrr, disturbing that there are still bugs after my long test with all its cases.  I fixed that and am now adding a permanent test and in the middle of confirming that the test fails (in a non-infinite-loop way: that's not a good thing in a test) if I put the size bug back.

Reply:
(If you're curious, the size check bugs are looking at the sizes of the wrong operands: registers instead of the memory operands.)

Reply:
On a related note: the following does not result in an infinite loop natively. I guess DR would handle this fine, since the size of load and store is the same (8 bytes)
```
label:
ldxp  w0, w1, [x2]
stxr  w3, x0, [x2]
cbnz  w3, label
```

Reply:
> (If you're curious, the size check bugs are looking at the sizes of the wrong operands: registers instead of the memory operands.)

Oh you mean the ordinal passed to `instr_get_dst`/`src` needed fixing?

Reply:
You've found another bug: the `ldxp w0,w1,[x2];stxr w3,x0,[x2]' fails both optimized and split because the load stores two separate 4-byte values into 2 separate 8-byte TLS slots, while the store checks against a single 8-byte value.  So many complexities with all these different opcodes.

I guess the solution is to have a pair load or store of 4-byte values combine them into a single TLS slot, and do a single 8-byte compare.  I've implemented this and added tests of both directions (ldxp;stxr and ldxr;stxp).